README.md

```diff
@@ -1 +1,62 @@
-# Readme
+# VS Verification Toolbox
+
+This module provides several useful tools for writing VS Code extensions that verify code.
+
+## Features
+
+### General Utilities
+
+The `util` package offers several convenient utilities and wrappers.
+
+#### File System Locations
+
+A `Location` represents a directory on the file system, providing a simple API for navigating through the hierarchy. Example usage:
+
+```typescript
+const root = new Location("/tmp/folder");
+root.basePath; // "/tmp/folder"
+root.enclosingFolder; // Location(/tmp)
+root.child("sub", "path"); // Location(/tmp/folder/sub/path)
+root.path("sub", "path"); // "/tmp/folder/sub/path"
+root.executable("run"); // unix-based: "/tmp/folder/run"; windows: "/tmp/folder/run.exe"
+await root.mkdir(); // ensures the folder exists
+await root.exists(); // true
+```
+
+#### Platform Checking
+
+`Platform` is a simple enum representing the three major platforms: `Linux`, `Windows`, `Mac`.
+
+It also provides a constant `currentPlatform` equal to the current platform, or `null` if it's not one of those three.
+
+#### Progress Reporting
+
+The `ProgressListener` type describes a callback for reporting progress from long-running asynchronous tasks. It takes the fraction of the task that is complete, as well as a user-facing description of the current step.
+
+Additionally, there is a very convenient function `withProgressInWindow` that runs an asynchronous task, forwarding the progress it reports (to the given `ProgressListener`) to the VS Code window as a notification with a nice progress bar.
+
+### Dependency Installation
+
+The `dependencies` package provides a way to define your dependencies declaratively and install them asynchronously.
+
+An example dependency might be defined as follows:
+
+```typescript
+const myDependency = new Dependency<"remote" | "local">(
+    "/path/to/install/folder",
```

Comment:
It's a bit confusing what this path means / is used for. As I understand it, it's a directory (under control of the plugin) to which remote dependencies are installed to. However, it has no impact on `LocalReference`, correct?

For me, it would be more natural to pass a destination path e.g. to the FileDownloader. What do you think?

Reply:
`LocalReference` is a `DependencyInstaller`, but it's exceptional in that its install action does nothing.

Currently the `DependencyInstaller` classes represent actions, but a different abstraction may be to rename them to represent sources (`RemoteZip`, `RemoteFile`, `LocalReference`, ...). Installing is an action that you can do on a source. `InstallerSequence` would then become just a function that composes progress listeners. I'm open to this.

---

In the example, I would recommend in a comment to use VS Code's `ExtensionContext.storagePath`, otherwise users might be tempted of manually crafting a path in the user's home depending on the OS.

Reply:
Good idea @fpoli; I will add that comment (or even just change the sample code directly).

The dependency system is by no means set in stone, but this is how it turned out after several iterations. There is somewhat of a tradeoff with composing installers: The current system defines very simple inputs and outputs (each a single file system location), but in return `InstallerSequence` can provide composition very easily. If instead installers were freer in how they define their inputs and outputs, they would have to manage their own composition, for which i can't think of a good way to implement something like `InstallerSequence`. I hope that makes sense!
It feels to me like the limitations of the current design are appropriate for the intended use of this system (i.e. downloading/installing dependencies).